plugins/devlink/fu-devlink-device.c

```diff
@@ -664,6 +664,106 @@ fu_devlink_device_cleanup(FuDevice *device,
 	return TRUE;
 }
 
+static void
+fu_devlink_device_add_json(FuDevice *device, JsonBuilder *builder, FwupdCodecFlags flags)
+{
+	FuDevlinkDevice *self = FU_DEVLINK_DEVICE(device);
+	GPtrArray *events = fu_device_get_events(device);
+
+	/* add device type identifier */
+	fwupd_codec_json_append(builder, "GType", "FuDevlinkDevice");
+
+	/* add devlink-specific properties for regular devices */
+	fwupd_codec_json_append(builder, "BusName", self->bus_name);
+	fwupd_codec_json_append(builder, "DevName", self->dev_name);
+
+	/* serialize recorded events */
+	if (events->len > 0) {
+		json_builder_set_member_name(builder, "Events");
+		json_builder_begin_array(builder);
+		for (guint i = 0; i < events->len; i++) {
+			FuDeviceEvent *event = g_ptr_array_index(events, i);
+
+			json_builder_begin_object(builder);
+			fwupd_codec_to_json(FWUPD_CODEC(event),
+					    builder,
+					    events->len > 1000 ? flags | FWUPD_CODEC_FLAG_COMPRESSED
+							       : flags);
+			json_builder_end_object(builder);
+		}
+		json_builder_end_array(builder);
+	}
+}
+
+static gboolean
+fu_devlink_device_from_json(FuDevice *device, JsonObject *json_object, GError **error)
+{
+	FuDevlinkDevice *self = FU_DEVLINK_DEVICE(device);
+	const gchar *bus_name;
+	const gchar *dev_name;
+	g_autofree gchar *device_id = NULL;
+
+	/* devlink-specific properties */
+	bus_name = json_object_get_string_member_with_default(json_object, "BusName", NULL);
+	dev_name = json_object_get_string_member_with_default(json_object, "DevName", NULL);
+
+	if (bus_name == NULL || dev_name == NULL) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_DATA,
+			    "BusName and DevName required for devlink device");
+		return FALSE;
+	}
+
+	g_free(self->bus_name);
+	g_free(self->dev_name);
+	self->bus_name = g_strdup(bus_name);
+	self->dev_name = g_strdup(dev_name);
```

Comment:
I think a `fu_devlink_device_set_bus_name()` helper here is probably a good idea